File: lib/python-sdk/common_grants_sdk/client/results.py
"""Typed result containers for the CommonGrants client.

``SearchResult`` / ``ListResult`` partition a batch response into the rows that
parsed (``items``) and the rows that did not (``errors``), so one malformed row
never fails the whole response. This is distinct from ``filter_info.errors``,
which carries filter-validation and server-reported errors.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Generic

from pydantic import BaseModel, ValidationError
from typing_extensions import TypeVar

from ..schemas.pydantic.pagination import PaginatedResultsInfo
from ..schemas.pydantic.responses.success import FilterInfo

ItemT = TypeVar("ItemT", bound=BaseModel)


@dataclass
class ParseFailure:
    """A single row that failed to parse into the target schema.

    ``index`` is the row's position in the response batch; ``raw`` is the
    unparsed row so a consumer can inspect or re-process it.
    """

    index: int
    message: str
    raw: dict[str, Any]


@dataclass
class ListResult(Generic[ItemT]):
    """A parsed list response: successfully parsed ``items`` plus per-row ``errors``."""

    items: list[ItemT]
    errors: list[ParseFailure]
    pagination_info: PaginatedResultsInfo


@dataclass
class SearchResult(Generic[ItemT]):
    """A parsed search response.

    ``items`` are the rows that parsed; ``errors`` are per-row ``ParseFailure``s.
    ``filter_info.errors`` is a separate channel carrying filter-validation and
    server-reported errors (fail-soft), not row parse failures.
    """

    items: list[ItemT]
    errors: list[ParseFailure]
    pagination_info: PaginatedResultsInfo
    filter_info: FilterInfo[Any]


def parse_batch(
    rows: list[dict[str, Any]], schema: type[ItemT]
) -> tuple[list[ItemT], list[ParseFailure]]:
    """Validate each row into ``schema``, collecting failures instead of raising.

    Returns ``(items, errors)``: ``items`` are the rows that validated, in order;
    ``errors`` are ``ParseFailure``s for the rows that did not. A single bad row
    never aborts the batch.
    """
    items: list[ItemT] = []
    errors: list[ParseFailure] = []
    for index, row in enumerate(rows):
        try:
            items.append(schema.model_validate(row))
        except ValidationError as exc:
            errors.append(ParseFailure(index=index, message=str(exc), raw=row))
    return items, errors

File: lib/ts-sdk/src/client/results.ts
/**
 * Per-row parse partitioning for list/search responses.
 *
 * `parseBatch` validates each response row against a schema and partitions the
 * outcome: valid rows land in `items`, failures in `errors` (each carrying its
 * row index and raw payload). One malformed row never aborts the batch unless
 * the caller opts into `"throw"`.
 */

import { z } from "zod";
import { Filtered, OppFilters, Paginated } from "../types";

/**
 * One response row that failed schema validation, with its position and raw payload.
 *
 * `index` is the row's position within its own page's raw rows, not the
 * aggregated `items` array.
 *
 * `raw` is the entire unvalidated row and may carry PII (e.g. in
 * `customFields`) — redact before logging, as with `TransformError.sourceValue`.
 */
export interface ParseFailure {
  index: number;
  raw: unknown;
  error: z.ZodError;
}

/** Thrown by `parseBatch` under `onParseError: "throw"`; carries the first failure's detail. */
export class BatchParseError extends Error {
  constructor(public readonly failure: ParseFailure) {
    super(`Row ${failure.index} failed to parse: ${failure.error.message}`);
    this.name = "BatchParseError";
  }
}

/** Row-parse failure handling: partition into `errors` (default) or throw on first failure. */
export type OnParseError = "collect" | "throw";

export function parseBatch<T>(
  schema: { safeParse(data: unknown): z.ZodSafeParseResult<T> },
  rows: unknown[],
  onParseError: OnParseError = "collect"
): { items: T[]; errors: ParseFailure[] } {
  const items: T[] = [];
  const errors: ParseFailure[] = [];
  rows.forEach((raw, index) => {
    const parsed = schema.safeParse(raw);
    if (parsed.success) {
      items.push(parsed.data);
      return;
    }
    const failure: ParseFailure = { index, raw, error: parsed.error };
    if (onParseError === "throw") throw new BatchParseError(failure);
    errors.push(failure);
  });
  return { items, errors };
}

/** Paginated result whose `items` are valid rows only; per-row failures land in `errors`. */
export type ListResult<T> = Paginated<T> & { errors: ParseFailure[] };

/**
 * Search result: filtered envelope plus per-row parse failures.
 *
 * `filterInfo` is optional because auto-pagination passes the server envelope
 * through unchanged, and a server may omit it.
 */
export type SearchResult<T, F = OppFilters> = Omit<Filtered<T, F>, "filterInfo"> & {
  filterInfo?: Filtered<T, F>["filterInfo"];
  errors: ParseFailure[];
};
